### DISCUSSION TITLE
Surreal syntax

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#9841] [Using MicroPython]
In my brief upython existence I've had more than my fair share of syntax errors.

Today, while trawling the web for a way to write tuples to files, I spotted a strange looking write syntax. I decide to try it on a rpi pico
```
>>> open('log', 'w').write('test test')
9
>>> with open('log', 'w')as f: f.read()
...
''
```
No syntax error, but no write either. Thinking that was maybe because it didn't close the file, I tried again
```
>>> open('log', 'w').write('test test'); open('log', 'w').close()
9
>>> with open('log', 'w')as f: f.read()
... 
'' 
```
Just curious what's happening here, the cmd is accepted but does'nt seem to do anything?

**Comment by LudoJoris** [comment]
I think the 'w' truncates the file. Try
>>> with open('log', 'r') as f: f.read()
or just leave the 'r' out, it's the default mode.

**Comment by jimmo** [comment]
> had more than my fair share of syntax errors.

@kjm1102 Not sure what this has to do with syntax (or syntax errors). Syntax errors are specifically when your program cannot be understood by the language. This is distinct to a runtime error (e.g. divide by zero or accessing an attribute that doesn't exist). Both of your snippets have correct syntax.

There are three issues here:
1. A files contents are not necessarily written in `write()`. Rather they are buffered up to either be written by a subsequent call to `flush()`, or when the file is closed.
2. In your second snippet, although you're closing _a_ file, you're not closing the same file object that you wrote to. Instead you need to do something like:

```py
f = open('log', 'w')
f.write('test test')
f.close()
```

Or better, have Python do the close automatically using a context manager (i.e. `with` statement)

```py
with open('log', 'w') as f:
  f.write('test test')
```

3. As @LudoJoris pointed out, if you want to read from a file, you must open it in read-mode (i.e. `'r'`). By specifying `'w'` it will truncate the file. Note that in CPython, it won't actually let you read from a file opened for writing. In MicroPython we allow it (I suspect because the extra code to detect this isn't worth the extra firmware size).

**Reply by kjm1102** [reply]
Victim of my own typo Jim, twice!
